File: uniube/1/licitacoes-hub/app/static/powerbi.js
/* Power BI / Dados Abertos PMU — removível */

let pbiDataset = "licitacoes";
let pbiAtualId = null;
let pbiPollTimer = null;

const PBI_DATASETS = {
  licitacoes: {
    label: "Licitações",
    cols: ["processo", "modalidade", "empresa", "ano_processo", "situacao", "valor_licitacao", "observador_nome", "objeto"],
    headers: ["Processo", "Modalidade", "Empresa", "Ano", "Situação", "Valor", "Observador", "Objeto"],
  },
  contratos: {
    label: "Contratos",
    cols: ["nr_contrato", "processo", "empresa", "ano_contrato", "nm_pessoa", "vr_inicial", "observador_nome", "ds_objeto_contrato"],
    headers: ["Nº contrato", "Processo", "Empresa", "Ano", "Fornecedor", "Valor", "Observador", "Objeto"],
  },
  gestores: {
    label: "Gestores e fiscais",
    cols: ["nr_contrato", "ds_papeis", "nm_pessoa_papel", "ds_orgao", "ano_contrato", "fornecedor", "observador_nome", "objeto_contrato"],
    headers: ["Nº contrato", "Papel", "Pessoa", "Órgão", "Ano", "Fornecedor", "Observador", "Objeto"],
  },
};

function pbiEsc(t) {
  const e = document.createElement("span");
  e.textContent = t ?? "—";
  return e.innerHTML;
}

function pbiFmtVal(row, col) {
  const v = row[col];
  if (v == null || v === "") return "—";
  if (col === "objeto" || col === "ds_objeto_contrato" || col === "objeto_contrato") {
    const s = String(v);
    return s.length > 90 ? `${s.slice(0, 90)}…` : s;
  }
  return v;
}

async function carregarPbiObservadores() {
  const obs = await api("/api/observadores?ativos=true");
  const sel = $("#edit-pbi-observador");
  if (!sel) return;
  sel.innerHTML = '<option value="">— Nenhum —</option>';
  obs.forEach((o) => {
    const opt = document.createElement("option");
    opt.value = o.id;
    opt.textContent = o.nome;
    sel.appendChild(opt);
  });
}

function pbiDatasetsSelecionados() {
  return [...document.querySelectorAll('#pbi-coleta-datasets input[name="dataset"]:checked')]
    .map((el) => el.value);
}

function pbiAnosSelecionados() {
  const raw = $("#pbi-coleta-anos")?.value || "";
  return raw.split(/[,;\s]+/).map((a) => parseInt(a.trim(), 10)).filter((a) => a >= 2000 && a <= 2100);
}

$("#form-pbi-coletar")?.addEventListener("submit", async (e) => {
  e.preventDefault();
  const datasets = pbiDatasetsSelecionados();
  const anos = pbiAnosSelecionados();
  if (!datasets.length) return alert("Selecione ao menos uma base.");
  if (datasets.some((d) => d !== "gestores") && !anos.length) {
    return alert("Informe ao menos um ano para licitações e/ou contratos.");
  }
  const btn = $("#btn-pbi-coletar");
  const logEl = $("#pbi-coleta-log");
  const resEl = $("#pbi-coleta-result");
  btn.disabled = true;
  logEl.classList.remove("hidden");
  resEl.classList.add("hidden");
  logEl.textContent = "Iniciando coleta…";
  try {
    await api("/api/powerbi/coletar", {
      method: "POST",
      body: JSON.stringify({ anos: anos.length ? anos : [new Date().getFullYear()], datasets }),
    });
    if (pbiPollTimer) clearInterval(pbiPollTimer);
    pbiPollTimer = setInterval(pollPbiColeta, 1200);
    pollPbiColeta();
  } catch (err) {
    logEl.textContent = err.message;
    btn.disabled = false;
  }
});

async function pollPbiColeta() {
  const st = await api("/api/powerbi/coletar/status");
  const logEl = $("#pbi-coleta-log");
  const resEl = $("#pbi-coleta-result");
  const btn = $("#btn-pbi-coletar");
  logEl.textContent = (st.log || []).join("\n") || "Aguardando…";
  logEl.scrollTop = logEl.scrollHeight;
  if (!st.running) {
    clearInterval(pbiPollTimer);
    pbiPollTimer = null;
    btn.disabled = false;
    if (st.resultado?.ok) {
      resEl.className = "result ok";
      resEl.textContent = "Coleta concluída com sucesso.";
      resEl.classList.remove("hidden");
    } else if (st.resultado) {
      resEl.className = "result err";
      resEl.textContent = st.resultado.erro || "Coleta finalizada com erros.";
      resEl.classList.remove("hidden");
    }
  }
}

function setPbiDataset(ds) {
  pbiDataset = ds;
  document.querySelectorAll(".pbi-tab").forEach((b) => {
    b.classList.toggle("active", b.dataset.dataset === ds);
  });
  document.querySelectorAll(".pbi-filtros-dataset").forEach((el) => {
    el.hidden = el.dataset.dataset !== ds;
  });
  buscarPbi();
}

async function carregarPbiFiltros() {
  const [empLic, sitLic, modLic, empCon, orgGest, papGest] = await Promise.all([
    api("/api/powerbi/licitacoes/empresas").catch(() => []),
    api("/api/powerbi/licitacoes/situacoes").catch(() => []),
    api("/api/powerbi/licitacoes/modalidades").catch(() => []),
    api("/api/powerbi/contratos/empresas").catch(() => []),
    api("/api/powerbi/gestores/orgaos").catch(() => []),
    api("/api/powerbi/gestores/papeis").catch(() => []),
  ]);
  const fill = (sel, items, allLabel) => {
    if (!sel) return;
    sel.innerHTML = `<option value="">${allLabel}</option>`;
    items.forEach((v) => {
      const o = document.createElement("option");
      o.value = v;
      o.textContent = v;
      sel.appendChild(o);
    });
  };
  fill($("#pbi-filtro-lic-empresa"), empLic, "Todas");
  fill($("#pbi-filtro-lic-situacao"), sitLic, "Todas");
  fill($("#pbi-filtro-lic-modalidade"), modLic, "Todas");
  fill($("#pbi-filtro-con-empresa"), empCon, "Todas");
  fill($("#pbi-filtro-gest-orgao"), orgGest, "Todos");
  fill($("#pbi-filtro-gest-papel"), papGest, "Todos");
}

function pbiQueryParams() {
  const p = new URLSearchParams();
  p.set("limit", "200");
  if (pbiDataset === "licitacoes") {
    const ano = $("#pbi-filtro-lic-ano")?.value;
    const empresa = $("#pbi-filtro-lic-empresa")?.value;
    const situacao = $("#pbi-filtro-lic-situacao")?.value;
    const modalidade = $("#pbi-filtro-lic-modalidade")?.value;
    const processo = $("#pbi-filtro-lic-processo")?.value;
    const texto = $("#pbi-filtro-lic-texto")?.value;
    if (ano) p.set("ano_processo", ano);
    if (empresa) p.set("empresa", empresa);
    if (situacao) p.set("situacao", situacao);
    if (modalidade) p.set("modalidade", modalidade);
    if (processo) p.set("processo", processo);
    if (texto) p.set("texto", texto);
  } else if (pbiDataset === "contratos") {
    const ano = $("#pbi-filtro-con-ano")?.value;
    const empresa = $("#pbi-filtro-con-empresa")?.value;
    const processo = $("#pbi-filtro-con-processo")?.value;
    const nr = $("#pbi-filtro-con-numero")?.value;
    const texto = $("#pbi-filtro-con-texto")?.value;
    if (ano) p.set("ano_contrato", ano);
    if (empresa) p.set("empresa", empresa);
    if (processo) p.set("processo", processo);
    if (nr) p.set("nr_contrato", nr);
    if (texto) p.set("texto", texto);
  } else {
    const ano = $("#pbi-filtro-gest-ano")?.value;
    const orgao = $("#pbi-filtro-gest-orgao")?.value;
    const papel = $("#pbi-filtro-gest-papel")?.value;
    const nr = $("#pbi-filtro-gest-numero")?.value;
    const texto = $("#pbi-filtro-gest-texto")?.value;
    if (ano) p.set("ano_contrato", ano);
    if (orgao) p.set("ds_orgao", orgao);
    if (papel) p.set("ds_papeis", papel);
    if (nr) p.set("nr_contrato", nr);
    if (texto) p.set("texto", texto);
  }
  return p;
}

async function buscarPbi() {
  const cfg = PBI_DATASETS[pbiDataset];
  const meta = $("#pbi-consulta-meta");
  const tbody = $("#pbi-tabela");
  if (!tbody) return;
  tbody.innerHTML = `<tr><td colspan="${cfg.headers.length}">Carregando…</td></tr>`;
  try {
    const data = await api(`/api/powerbi/${pbiDataset}?${pbiQueryParams()}`);
    meta.textContent = `${data.total.toLocaleString("pt-BR")} registro(s) — ${cfg.label}`;
    if (!data.items.length) {
      tbody.innerHTML = `<tr><td colspan="${cfg.headers.length}">Nenhum registro.</td></tr>`;
      return;
    }
    tbody.innerHTML = data.items.map((row) => `
      <tr data-id="${row.id}" class="clickable">
        ${cfg.cols.map((c) => `<td>${pbiEsc(pbiFmtVal(row, c))}</td>`).join("")}
      </tr>`).join("");
    tbody.querySelectorAll("tr.clickable").forEach((tr) => {
      tr.addEventListener("click", () => abrirPbiModal(parseInt(tr.dataset.id, 10)));
    });
  } catch (err) {
    tbody.innerHTML = `<tr><td colspan="${cfg.headers.length}">${pbiEsc(err.message)}</td></tr>`;
  }
}

async function abrirPbiModal(id) {
  pbiAtualId = id;
  const row = await api(`/api/powerbi/${pbiDataset}/${id}`);
  $("#modal-pbi-titulo").textContent = `${PBI_DATASETS[pbiDataset].label} #${id}`;
  const grid = $("#modal-pbi-info");
  const skip = new Set(["id", "observador_id", "observador_nome", "coletado_em", "dados_csv_json"]);
  const labels = {
    ano_processo: "ANOPROCESSO", chave: "CHAVE", ds_habilitacao: "DSHABILITACAO",
    dt_abertura: "DTABERTURA", dt_habilitacao: "DTHABILITACAO", dt_homologacao: "DTHOMOLOGACAO",
    dt_julgamento: "DTJULGAMENTO", valor_licitacao: "VALORLICITACAO", fonte_ano_coleta: "Ano coleta",
    ano_contrato: "ANOCONTRATO", ds_objeto_contrato: "DSOBJETOCONTRATO", dt_assinatura: "DTASSINATURA",
    nm_pessoa: "NMPESSOA", nr_aditivo: "NRADITIVO", nr_contrato: "NRCONTRATO", nr_parcela: "NRPARCELA",
    vr_inicial: "VRINICIAL", ds_orgao: "DSORGAO", ds_papeis: "DSPAPEIS", dt_fim: "DTFIM",
    dt_inicio: "DTINICIO", fornecedor: "FORNECEDOR", nm_pessoa_papel: "NMPESSOAPAPEL",
    objeto_contrato: "OBJETOCONTRATO", processo: "PROCESSO", modalidade: "MODALIDADE",
    objeto: "OBJETO", solicitante: "SOLICITANTE", empresa: "EMPRESA", situacao: "SITUACAO",
  };
  grid.innerHTML = Object.entries(row)
    .filter(([k, v]) => !skip.has(k) && v != null && v !== "")
    .map(([k, v]) => `<dt>${pbiEsc(labels[k] || k)}</dt><dd>${pbiEsc(String(v))}</dd>`)
    .join("");
  const obj = row.objeto || row.ds_objeto_contrato || row.objeto_contrato;
  const objEl = $("#modal-pbi-objeto");
  if (objEl) {
    objEl.textContent = obj || "—";
    objEl.closest(".detail-panel")?.toggleAttribute("hidden", !obj);
  }
  await carregarPbiObservadores();
  $("#edit-pbi-observador").value = row.observador_id || "";
  $("#modal-pbi").showModal();
}

$("#form-pbi-filtros")?.addEventListener("submit", (e) => {
  e.preventDefault();
  buscarPbi();
});

$("#btn-pbi-limpar")?.addEventListener("click", () => {
  document.querySelectorAll(".pbi-filtros-dataset:not([hidden]) input, .pbi-filtros-dataset:not([hidden]) select")
    .forEach((el) => { if (el.tagName === "SELECT") el.selectedIndex = 0; else el.value = ""; });
  buscarPbi();
});

$("#form-pbi-obs")?.addEventListener("submit", async (e) => {
  e.preventDefault();
  if (!pbiAtualId) return;
  const obs = $("#edit-pbi-observador").value;
  try {
    await api(`/api/powerbi/${pbiDataset}/${pbiAtualId}`, {
      method: "PATCH",
      body: JSON.stringify({ observador_id: obs ? parseInt(obs, 10) : 0 }),
    });
    $("#modal-pbi").close();
    buscarPbi();
  } catch (err) {
    alert(err.message);
  }
});

$("#modal-pbi-fechar")?.addEventListener("click", () => $("#modal-pbi").close());

document.querySelectorAll(".pbi-tab").forEach((btn) => {
  btn.addEventListener("click", () => setPbiDataset(btn.dataset.dataset));
});

function atualizarPbiTabelaHead() {
  const cfg = PBI_DATASETS[pbiDataset];
  const head = $("#pbi-tabela-head");
  if (!head) return;
  head.innerHTML = `<tr>${cfg.headers.map((h) => `<th>${h}</th>`).join("")}</tr>`;
}

$$(".nav-btn").forEach((btn) => {
  btn.addEventListener("click", () => {
    if (btn.dataset.page === "powerbi-consultar") {
      atualizarPbiTabelaHead();
      carregarPbiFiltros();
      buscarPbi();
    }
  });
});

atualizarPbiTabelaHead();

File: uniube/1/licitacoes-hub/app/powerbi.py
"""API Power BI / Dados Abertos PMU — módulo isolado (removível)."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy import func, or_, select
from sqlalchemy.orm import Session, selectinload

from app.config import POWERBI_DATASETS, POWERBI_PANEL_URL
from app.database import (
    PowerBiContrato,
    PowerBiGestorFiscal,
    PowerBiLicitacao,
    SessionLocal,
    get_db,
)
from app.powerbi_coletor import coletar as coletar_powerbi

router = APIRouter(tags=["powerbi"])

_coleta_status: dict[str, Any] = {"running": False, "log": [], "resultado": None}


class PowerBiColetaRequest(BaseModel):
    anos: list[int] = Field(min_length=1)
    datasets: list[str] = Field(
        default=["licitacoes", "contratos", "gestores"],
        description="licitacoes | contratos | gestores",
    )


class PowerBiLicitacaoOut(BaseModel):
    id: int
    ano_processo: int
    chave: str | None = None
    processo: str
    modalidade: str
    objeto: str | None = None
    solicitante: str | None = None
    empresa: str
    situacao: str | None = None
    ds_habilitacao: str | None = None
    dt_abertura: str | None = None
    dt_habilitacao: str | None = None
    dt_julgamento: str | None = None
    dt_homologacao: str | None = None
    valor_licitacao: str | None = None
    fonte_ano_coleta: int
    observador_id: int | None = None
    observador_nome: str | None = None
    coletado_em: datetime | None = None

    model_config = {"from_attributes": True}


class PowerBiContratoOut(BaseModel):
    id: int
    ano_contrato: int
    ano_processo: int | None = None
    ds_objeto_contrato: str | None = None
    dt_assinatura: str | None = None
    empresa: str
    nm_pessoa: str | None = None
    nr_aditivo: str | None = None
    nr_contrato: str
    nr_parcela: str | None = None
    processo: str | None = None
    vr_inicial: str | None = None
    fonte_ano_coleta: int
    observador_id: int | None = None
    observador_nome: str | None = None
    coletado_em: datetime | None = None

    model_config = {"from_attributes": True}


class PowerBiGestorOut(BaseModel):
    id: int
    ano_contrato: int
    ds_orgao: str | None = None
    ds_papeis: str | None = None
    dt_assinatura: str | None = None
    dt_fim: str | None = None
    dt_inicio: str | None = None
    fornecedor: str | None = None
    nm_pessoa_papel: str
    nr_contrato: str
    objeto_contrato: str | None = None
    observador_id: int | None = None
    observador_nome: str | None = None
    coletado_em: datetime | None = None

    model_config = {"from_attributes": True}


class PowerBiObservadorUpdate(BaseModel):
    observador_id: int | None = None


def _lic_out(row: PowerBiLicitacao) -> PowerBiLicitacaoOut:
    data = {c: getattr(row, c) for c in PowerBiLicitacaoOut.model_fields if c != "observador_nome"}
    data["observador_nome"] = row.observador.nome if row.observador else None
    return PowerBiLicitacaoOut(**data)


def _con_out(row: PowerBiContrato) -> PowerBiContratoOut:
    data = {c: getattr(row, c) for c in PowerBiContratoOut.model_fields if c != "observador_nome"}
    data["observador_nome"] = row.observador.nome if row.observador else None
    return PowerBiContratoOut(**data)


def _gest_out(row: PowerBiGestorFiscal) -> PowerBiGestorOut:
    data = {c: getattr(row, c) for c in PowerBiGestorOut.model_fields if c != "observador_nome"}
    data["observador_nome"] = row.observador.nome if row.observador else None
    return PowerBiGestorOut(**data)


def _run_coleta(anos: list[int], datasets: list[str]) -> None:
    db = SessionLocal()
    try:
        resultado = coletar_powerbi(
            db,
            anos,
            datasets,
            on_log=lambda m: _coleta_status["log"].append(m),
        )
        _coleta_status["resultado"] = {"ok": True, **resultado}
    except Exception as exc:
        _coleta_status["log"].append(f"Erro: {exc}")
        _coleta_status["resultado"] = {"ok": False, "erro": str(exc)}
    finally:
        db.close()
        _coleta_status["running"] = False


@router.get("/api/powerbi/fontes")
def fontes():
    return {
        "painel_url": POWERBI_PANEL_URL,
        "datasets": [
            {"id": k, "nome": v["nome"], "endpoint": v["endpoint"]}
            for k, v in POWERBI_DATASETS.items()
        ],
    }


@router.get("/api/powerbi/stats")
def stats(db: Session = Depends(get_db)):
    return {
        "licitacoes": db.scalar(select(func.count()).select_from(PowerBiLicitacao)) or 0,
        "contratos": db.scalar(select(func.count()).select_from(PowerBiContrato)) or 0,
        "gestores": db.scalar(select(func.count()).select_from(PowerBiGestorFiscal)) or 0,
    }


@router.post("/api/powerbi/coletar")
def iniciar_coleta(req: PowerBiColetaRequest, bg: BackgroundTasks):
    invalidos = [d for d in req.datasets if d not in POWERBI_DATASETS]
    if invalidos:
        raise HTTPException(400, f"Datasets inválidos: {', '.join(invalidos)}")
    precisa_ano = any(d in req.datasets for d in ("licitacoes", "contratos"))
    if precisa_ano and not req.anos:
        raise HTTPException(400, "Informe ao menos um ano para licitações/contratos")
    if _coleta_status["running"]:
        raise HTTPException(409, "Já existe uma coleta Power BI em andamento")

    _coleta_status.update(running=True, log=[], resultado=None)
    bg.add_task(_run_coleta, req.anos, req.datasets)
    return {"status": "iniciada", "anos": req.anos, "datasets": req.datasets}


@router.get("/api/powerbi/coletar/status")
def status_coleta():
    return _coleta_status


@router.get("/api/powerbi/licitacoes/empresas")
def licitacoes_empresas(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiLicitacao.empresa)
        .distinct()
        .where(PowerBiLicitacao.empresa.isnot(None), PowerBiLicitacao.empresa != "")
        .order_by(PowerBiLicitacao.empresa)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/licitacoes/situacoes")
def licitacoes_situacoes(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiLicitacao.situacao)
        .distinct()
        .where(PowerBiLicitacao.situacao.isnot(None), PowerBiLicitacao.situacao != "")
        .order_by(PowerBiLicitacao.situacao)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/licitacoes/modalidades")
def licitacoes_modalidades(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiLicitacao.modalidade)
        .distinct()
        .where(PowerBiLicitacao.modalidade.isnot(None), PowerBiLicitacao.modalidade != "")
        .order_by(PowerBiLicitacao.modalidade)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/contratos/empresas")
def contratos_empresas(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiContrato.empresa)
        .distinct()
        .where(PowerBiContrato.empresa.isnot(None), PowerBiContrato.empresa != "")
        .order_by(PowerBiContrato.empresa)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/gestores/orgaos")
def gestores_orgaos(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiGestorFiscal.ds_orgao)
        .distinct()
        .where(PowerBiGestorFiscal.ds_orgao.isnot(None), PowerBiGestorFiscal.ds_orgao != "")
        .order_by(PowerBiGestorFiscal.ds_orgao)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/gestores/papeis")
def gestores_papeis(db: Session = Depends(get_db)):
    rows = db.scalars(
        select(PowerBiGestorFiscal.ds_papeis)
        .distinct()
        .where(PowerBiGestorFiscal.ds_papeis.isnot(None), PowerBiGestorFiscal.ds_papeis != "")
        .order_by(PowerBiGestorFiscal.ds_papeis)
    ).all()
    return [r for r in rows if r]


@router.get("/api/powerbi/licitacoes")
def listar_licitacoes(
    db: Session = Depends(get_db),
    ano_processo: int | None = None,
    fonte_ano: int | None = None,
    empresa: str | None = None,
    situacao: str | None = None,
    modalidade: str | None = None,
    processo: str | None = None,
    texto: str | None = None,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    stmt = select(PowerBiLicitacao).options(selectinload(PowerBiLicitacao.observador))
    count_stmt = select(func.count()).select_from(PowerBiLicitacao)

    if ano_processo:
        stmt = stmt.where(PowerBiLicitacao.ano_processo == ano_processo)
        count_stmt = count_stmt.where(PowerBiLicitacao.ano_processo == ano_processo)
    if fonte_ano:
        stmt = stmt.where(PowerBiLicitacao.fonte_ano_coleta == fonte_ano)
        count_stmt = count_stmt.where(PowerBiLicitacao.fonte_ano_coleta == fonte_ano)
    if empresa:
        stmt = stmt.where(PowerBiLicitacao.empresa == empresa)
        count_stmt = count_stmt.where(PowerBiLicitacao.empresa == empresa)
    if situacao:
        stmt = stmt.where(PowerBiLicitacao.situacao == situacao)
        count_stmt = count_stmt.where(PowerBiLicitacao.situacao == situacao)
    if modalidade:
        stmt = stmt.where(PowerBiLicitacao.modalidade == modalidade)
        count_stmt = count_stmt.where(PowerBiLicitacao.modalidade == modalidade)
    if processo:
        p = f"%{processo.strip()}%"
        stmt = stmt.where(PowerBiLicitacao.processo.ilike(p))
        count_stmt = count_stmt.where(PowerBiLicitacao.processo.ilike(p))
    if texto:
        t = f"%{texto.strip()}%"
        filtro = or_(
            PowerBiLicitacao.objeto.ilike(t),
            PowerBiLicitacao.solicitante.ilike(t),
            PowerBiLicitacao.empresa.ilike(t),
            PowerBiLicitacao.modalidade.ilike(t),
        )
        stmt = stmt.where(filtro)
        count_stmt = count_stmt.where(filtro)

    total = db.scalar(count_stmt) or 0
    rows = db.scalars(
        stmt.order_by(
            PowerBiLicitacao.ano_processo.desc(),
            PowerBiLicitacao.empresa,
            PowerBiLicitacao.processo,
        )
        .offset(offset)
        .limit(limit)
    ).all()
    return {"total": total, "items": [_lic_out(r) for r in rows]}


@router.get("/api/powerbi/licitacoes/{lid}", response_model=PowerBiLicitacaoOut)
def obter_licitacao(lid: int, db: Session = Depends(get_db)):
    row = db.scalar(
        select(PowerBiLicitacao)
        .options(selectinload(PowerBiLicitacao.observador))
        .where(PowerBiLicitacao.id == lid)
    )
    if not row:
        raise HTTPException(404, "Não encontrada")
    return _lic_out(row)


@router.get("/api/powerbi/contratos")
def listar_contratos(
    db: Session = Depends(get_db),
    ano_contrato: int | None = None,
    ano_processo: int | None = None,
    fonte_ano: int | None = None,
    empresa: str | None = None,
    processo: str | None = None,
    nr_contrato: str | None = None,
    texto: str | None = None,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    stmt = select(PowerBiContrato).options(selectinload(PowerBiContrato.observador))
    count_stmt = select(func.count()).select_from(PowerBiContrato)

    if ano_contrato:
        stmt = stmt.where(PowerBiContrato.ano_contrato == ano_contrato)
        count_stmt = count_stmt.where(PowerBiContrato.ano_contrato == ano_contrato)
    if ano_processo:
        stmt = stmt.where(PowerBiContrato.ano_processo == ano_processo)
        count_stmt = count_stmt.where(PowerBiContrato.ano_processo == ano_processo)
    if fonte_ano:
        stmt = stmt.where(PowerBiContrato.fonte_ano_coleta == fonte_ano)
        count_stmt = count_stmt.where(PowerBiContrato.fonte_ano_coleta == fonte_ano)
    if empresa:
        stmt = stmt.where(PowerBiContrato.empresa == empresa)
        count_stmt = count_stmt.where(PowerBiContrato.empresa == empresa)
    if processo:
        p = f"%{processo.strip()}%"
        stmt = stmt.where(PowerBiContrato.processo.ilike(p))
        count_stmt = count_stmt.where(PowerBiContrato.processo.ilike(p))
    if nr_contrato:
        n = f"%{nr_contrato.strip()}%"
        stmt = stmt.where(PowerBiContrato.nr_contrato.ilike(n))
        count_stmt = count_stmt.where(PowerBiContrato.nr_contrato.ilike(n))
    if texto:
        t = f"%{texto.strip()}%"
        filtro = or_(
            PowerBiContrato.ds_objeto_contrato.ilike(t),
            PowerBiContrato.nm_pessoa.ilike(t),
            PowerBiContrato.empresa.ilike(t),
        )
        stmt = stmt.where(filtro)
        count_stmt = count_stmt.where(filtro)

    total = db.scalar(count_stmt) or 0
    rows = db.scalars(
        stmt.order_by(
            PowerBiContrato.ano_contrato.desc(),
            PowerBiContrato.nr_contrato,
        )
        .offset(offset)
        .limit(limit)
    ).all()
    return {"total": total, "items": [_con_out(r) for r in rows]}


@router.get("/api/powerbi/contratos/{cid}", response_model=PowerBiContratoOut)
def obter_contrato(cid: int, db: Session = Depends(get_db)):
    row = db.scalar(
        select(PowerBiContrato)
        .options(selectinload(PowerBiContrato.observador))
        .where(PowerBiContrato.id == cid)
    )
    if not row:
        raise HTTPException(404, "Não encontrado")
    return _con_out(row)


@router.get("/api/powerbi/gestores")
def listar_gestores(
    db: Session = Depends(get_db),
    ano_contrato: int | None = None,
    ds_orgao: str | None = None,
    ds_papeis: str | None = None,
    nr_contrato: str | None = None,
    fornecedor: str | None = None,
    texto: str | None = None,
    limit: int = Query(100, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    stmt = select(PowerBiGestorFiscal).options(selectinload(PowerBiGestorFiscal.observador))
    count_stmt = select(func.count()).select_from(PowerBiGestorFiscal)

    if ano_contrato:
        stmt = stmt.where(PowerBiGestorFiscal.ano_contrato == ano_contrato)
        count_stmt = count_stmt.where(PowerBiGestorFiscal.ano_contrato == ano_contrato)
    if ds_orgao:
        stmt = stmt.where(PowerBiGestorFiscal.ds_orgao == ds_orgao)
        count_stmt = count_stmt.where(PowerBiGestorFiscal.ds_orgao == ds_orgao)
    if ds_papeis:
        stmt = stmt.where(PowerBiGestorFiscal.ds_papeis == ds_papeis)
        count_stmt = count_stmt.where(PowerBiGestorFiscal.ds_papeis == ds_papeis)
    if nr_contrato:
        n = f"%{nr_contrato.strip()}%"
        stmt = stmt.where(PowerBiGestorFiscal.nr_contrato.ilike(n))
        count_stmt = count_stmt.where(PowerBiGestorFiscal.nr_contrato.ilike(n))
    if fornecedor:
        f = f"%{fornecedor.strip()}%"
        stmt = stmt.where(PowerBiGestorFiscal.fornecedor.ilike(f))
        count_stmt = count_stmt.where(PowerBiGestorFiscal.fornecedor.ilike(f))
    if texto:
        t = f"%{texto.strip()}%"
        filtro = or_(
            PowerBiGestorFiscal.nm_pessoa_papel.ilike(t),
            PowerBiGestorFiscal.objeto_contrato.ilike(t),
            PowerBiGestorFiscal.fornecedor.ilike(t),
        )
        stmt = stmt.where(filtro)
        count_stmt = count_stmt.where(filtro)

    total = db.scalar(count_stmt) or 0
    rows = db.scalars(
        stmt.order_by(
            PowerBiGestorFiscal.ano_contrato.desc(),
            PowerBiGestorFiscal.nr_contrato,
            PowerBiGestorFiscal.nm_pessoa_papel,
        )
        .offset(offset)
        .limit(limit)
    ).all()
    return {"total": total, "items": [_gest_out(r) for r in rows]}


@router.get("/api/powerbi/gestores/{gid}", response_model=PowerBiGestorOut)
def obter_gestor(gid: int, db: Session = Depends(get_db)):
    row = db.scalar(
        select(PowerBiGestorFiscal)
        .options(selectinload(PowerBiGestorFiscal.observador))
        .where(PowerBiGestorFiscal.id == gid)
    )
    if not row:
        raise HTTPException(404, "Não encontrado")
    return _gest_out(row)


@router.patch("/api/powerbi/licitacoes/{lid}", response_model=PowerBiLicitacaoOut)
def atualizar_licitacao_pbi(lid: int, payload: PowerBiObservadorUpdate, db: Session = Depends(get_db)):
    from app.database import Observador

    row = db.scalar(
        select(PowerBiLicitacao)
        .options(selectinload(PowerBiLicitacao.observador))
        .where(PowerBiLicitacao.id == lid)
    )
    if not row:
        raise HTTPException(404, "Não encontrada")
    if payload.observador_id is not None:
        if payload.observador_id == 0:
            row.observador_id = None
        else:
            obs = db.get(Observador, payload.observador_id)
            if not obs or not obs.ativo:
                raise HTTPException(400, "Observador inválido ou inativo")
            row.observador_id = payload.observador_id
    db.commit()
    db.refresh(row)
    return _lic_out(row)


@router.patch("/api/powerbi/contratos/{cid}", response_model=PowerBiContratoOut)
def atualizar_contrato_pbi(cid: int, payload: PowerBiObservadorUpdate, db: Session = Depends(get_db)):
    from app.database import Observador

    row = db.scalar(
        select(PowerBiContrato)
        .options(selectinload(PowerBiContrato.observador))
        .where(PowerBiContrato.id == cid)
    )
    if not row:
        raise HTTPException(404, "Não encontrado")
    if payload.observador_id is not None:
        if payload.observador_id == 0:
            row.observador_id = None
        else:
            obs = db.get(Observador, payload.observador_id)
            if not obs or not obs.ativo:
                raise HTTPException(400, "Observador inválido ou inativo")
            row.observador_id = payload.observador_id
    db.commit()
    db.refresh(row)
    return _con_out(row)


@router.patch("/api/powerbi/gestores/{gid}", response_model=PowerBiGestorOut)
def atualizar_gestor_pbi(gid: int, payload: PowerBiObservadorUpdate, db: Session = Depends(get_db)):
    from app.database import Observador

    row = db.scalar(
        select(PowerBiGestorFiscal)
        .options(selectinload(PowerBiGestorFiscal.observador))
        .where(PowerBiGestorFiscal.id == gid)
    )
    if not row:
        raise HTTPException(404, "Não encontrado")
    if payload.observador_id is not None:
        if payload.observador_id == 0:
            row.observador_id = None
        else:
            obs = db.get(Observador, payload.observador_id)
            if not obs or not obs.ativo:
                raise HTTPException(400, "Observador inválido ou inativo")
            row.observador_id = payload.observador_id
    db.commit()
    db.refresh(row)
    return _gest_out(row)
